crates/sui-framework/packages/bridge/sources/limiter.move

```diff
@@ -0,0 +1,192 @@
+        last_recorded_hour: u64,
+        per_hour_amounts: vector<u64>,
+        total_amount: u64
+    }
+
+    public fun new(transfer_limits: VecMap<BridgeRoute, u64>): TransferLimiter {
+        TransferLimiter {
+            transfer_limits,
+            notional_values: vec_map::empty(),
+            transfer_records: vec_map::empty()
+        }
+    }
+
+    // Current hour since unix epoch
+    fun current_hour_since_epoch(clock: &Clock): u64 {
+        clock::timestamp_ms(clock) / 3600000
+    }
+
+    public fun check_and_record_transfer<T>(
+        clock: & Clock,
+        self: &mut TransferLimiter,
+        route: BridgeRoute,
+        amount: u64
+    ) {
+        // Create record for route if not exists
+        if (!vec_map::contains(&self.transfer_records, &route)) {
+            vec_map::insert(&mut self.transfer_records, route, TransferRecord {
+                last_recorded_hour: 0,
+                per_hour_amounts: vector[],
+                total_amount: 0
+            })
+        };
+        let record = vec_map::get_mut(&mut self.transfer_records, &route);
+        let current_hour_since_epoch = current_hour_since_epoch(clock);
+
+        // First clean up old transfer histories
+        let i = vector::length(&record.per_hour_amounts);
+        while (i > 0 && record.last_recorded_hour < (current_hour_since_epoch - 24 + i)) {
+            record.total_amount = record.total_amount - vector::remove(&mut record.per_hour_amounts, 0);
+            i = i - 1;
+        };
+        // Backfill missing hours
+        while (record.last_recorded_hour < current_hour_since_epoch) {
+            // only backfill up to 24 hours
+            if (record.last_recorded_hour < current_hour_since_epoch - 23) {
+                record.last_recorded_hour = current_hour_since_epoch - 23;
+            }else {
+                record.last_recorded_hour = record.last_recorded_hour + 1;
+            };
+            vector::push_back(&mut record.per_hour_amounts, 0);
+        };
+
+        // Get limit for the route
+        let route_limit = if (vec_map::contains(&self.transfer_limits, &route)) {
+            *vec_map::get(&self.transfer_limits, &route)
+        } else {
+            DEFAULT_TRANSFER_LIMIT
+        };
+
+        // Compute notional amount
+        let coin_type = type_name::get<T>();
+        let notional_amount = *vec_map::get(&self.notional_values, &coin_type) * amount;
+        let notional_amount = notional_amount / pow(10, treasury::token_decimals<T>());
+
+        // Check if transfer amount exceed limit
+        assert!(record.total_amount + notional_amount <= route_limit, ETransferAmountExceedLimit);
+
+        // Record transfer value
+        let new_amount = vector::pop_back(&mut record.per_hour_amounts) + notional_amount;
+        vector::push_back(&mut record.per_hour_amounts, new_amount);
+        record.total_amount = record.total_amount + notional_amount;
+    }
+
+    #[test]
```

Comment:
can we have a unit test for two different routes that not interfering with each other?